[SofaPython3] proposal of feature to share.

damienmarchal opened the pull request:
The SofaPython plugin is now outdated and based on python2. 

As python3 already breaks the API, updating the SofaPython plugin for python3 is an opportunity to also refactor the whole binding code. 

Stuff I would like to have.
Pythonic way to manipulate data
```python
  c = object1.addObject("MechanicalObject", position=[0,0,0]*1000)
  len(c.position)
  for i in c.position:
        print i
  c.position += [1,0,0]
  c.position[2] += [1,0,0]
  c.position[2,0] += 2
  c.position[1:10] = [1,0,0] 
  ...
```

Numpy exposure of data
```python
  c = object1.addObject("MechanicalObject", position=[0,0,0]*1000)

  ## Read access (copy-less) using toarray()
  p = c.position.toarray()
  o = numpy.ones((len(c.position), 3)))  
  j = p + o

  ## Write access (copy-less), the data counter is updated after the 'with' block
  with c.position.writeable() as w:
          o = numpy.ones((len(c.position), 3)))  
          w += o 
          w *= 2.0
```

Wrap around types (copy-less) to typify the data fields on demand
```python
# syntax 1
c = o.showColor.as(RGBAColor)
# syntax 2
c = RGBAColor(o.showColor)
c.setFromHSC(1,2,3)
c.setFromHSV(1,2,3)
```

Implement your own ForceField in python:
```python
class MyForceField(Sofa.Core.BaseForceField):
    def __init__(self, *args, **kwargs):
        Sofa.Core.BaseForceField.__init__(self, *args, **kwargs)
    
    def init(self):
        self.refx = self.position.toarray().copy()

    def addForce(self, m, f, x, v):
        n = numpy.ones((len(self.position),1))
        f =  (self.refx-x) * 2 + n
```
    
To test the binding you need to de-activate SofaPython and activate SofaPython3. 
Then
- runSimu.py to launch a sofa simulation from the command line eg:   
  python3 runSimu.py mySimulation.py
- or runSofa. 

General guidelines:
- Fix existing API consistency issues
- Improve the performance of the binding by supporting supporting the different python protocols (sequence, buffer+memoryview, copy-less mechanism).
- unified data access and API.
- implementation is currently done with pybind11 and cython.
- modular design.

About the tools we use: 
- pybind because it is very consistent with our code base 
- cython becaues it is much more powerful and general and so we can accelerate python code by un-warping the warped object in other part of the code base. 

______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

damienmarchal committed:
[cmake] ADD latestest version of FindCython.cmake & sofa_add_python_module in SofaMacros

(To avoid PR dependencies).

damienmarchal committed:
[modules/SofaGeometry] ADD the c++ module which contains trivial geometrical object/operations.

The module contains Plane,Ray and some constants.

damienmarchal committed:
[modules/SofaGeometry] ADD a dedicated binding to the SofaGeometry module.

The binding is made with CYTHON.

damienmarchal committed:
[modules] ADD SofaGeometry in the main CMakeLists.txt and update internal one to make it compile.

damienmarchal committed:
[modules/SofaGeometry] ADD tests subdirectory in CMakeLists.txt

damienmarchal committed:
[SofaKernel] FIX a bug in sofa_add_python_module (from SofaMacros.cmake)

damienmarchal committed:
[SofaGeometry] Updates the CMakeLists.txt to activate the tests & the binding.

damienmarchal committed:
[SofaGeometry] ADD default binding for SofaGeometry

I'm not sure having these files on the repo is needed but I found that usefull
for the ones that does not have cython installed (eg: the CI happy).

damienmarchal committed:
[SofaGeometry] FIX Adds missing SOFAGEOMETRY_API which prevent linking on windows.

damienmarchal committed:
[SofaGeometry] Remove the .cpp automatically generated from cython.

We now have to install cython to generate this binding.

damienmarchal committed:
[SofaGeometry] UPDATE SofaGeometry to force the use of Cython.

damienmarchal committed:
[CMakeLists] FIX CMakeLists to handle modules & tests

damienmarchal committed:
[SofaGeometry] ADD tests in SofaGeometry and improve implementation.

damienmarchal committed:
[SofaKernel/SofaMacro] Fix sofa_add_python_module that was broken

damienmarchal committed:
[SofaPython] Sofa.Data refactoring.

damienmarchal committed:
[SofaPython3] ADD a new module for SofaPython3

damienmarchal committed:
Merge branch 'sofa_addSofaGeometryModule' into sofa_refactorSofaPython

# Conflicts:
#	modules/CMakeLists.txt

damienmarchal committed:
[Update] FindCython to have a directory parameter.

damienmarchal committed:
[SofaGeometry] Add Package.

damienmarchal committed:
[SofaPython3]

damienmarchal committed:
[SofaPython3] Experimental layout

damienmarchal committed:
[pybind] For Base.

damienmarchal committed:
[pybind] Base

damienmarchal committed:
[pybind]

damienmarchal committed:
[SofPython3] pybind.

damienmarchal committed:
Merge branch 'master' of github.com:sofa-framework/sofa into sofa_refactorSofaPython

damienmarchal committed:
[SofaKernel] SofaMacro

The pybind module is now a real shared library so we can link it to other target to
establish the dependencies.

damienmarchal committed:
[SofaPython3/binding/Sofa] begin refactor

The idea is to investigate now how inter-module can be managed.

damienmarchal committed:
[SofaPython3/bindings/SofaRuntime] create SofaRuntime.  (WIP)

damienmarchal committed:
[SofaRuntime] Add new bug.

damienmarchal committed:
[SofaPython3] Sofa module layout...

damienmarchal committed:
[SofaPython3] Add PythonTest class (NOT WORKING YET)

damienmarchal committed:
[SofaPython3] Update CMAKELISTS.

damienmarchal committed:
[SofaPython3] Fall back to python3.

damienmarchal committed:
[SofaPython3] Adds a minimal tests to show problems in the binding.

damienmarchal committed:
[SofaPython3] Add TODO.

marques-bruno committed:
[pybind] Add Bindings_Simulation

damienmarchal committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into sofa_refactorSofaPython

damienmarchal committed:
Merge branch 'sofa_refactorSofaPython' of github.com:SofaDefrost/sofa into sofa_refactorSofaPython

damienmarchal committed:
[SofaPython] WIP

damienmarchal committed:
[SofaKernel] Add str() & importPlugin in simpleapi.

damienmarchal committed:
[SofaPython3] WIP/changé le testé pour créer les objets.

marques-bruno committed:
[pybind] ADD SofaTypes module

damienmarchal committed:
[SofaKernel/Simulation.cpp] remove assert.

I removed the assert because it is inconsistent API. In which the initialization
of the smart pointer to nullptr is clashing with the assert saying that the smartpointer
cannot be NULL.

So:
 - either the function is supposed to work with a nullptr scene
   and in that case the assert is useless. So nullity of the pointer is valid for
   certain function.
- or, the function cannot work with a nullptr and in that case using an assert
  is not the right way to implement functional code that should report message to
  caller's.

damienmarchal committed:
[SofaKernel] Remove un-needed dependency to Node.h

damienmarchal committed:
[SofaKernel] CLEAN Simulation static API

By:
 - remove the static Simulation in the Simulation object
   (it is not needed to force the simulation object to have a static entry)
 - Change the setSimulation signature to make it clear that a SPtr is expected thus delegating
   the delete of the object to the SPtr
 - Remove an assert to allow nullptr to be manipulated to be consistent with the fact that the Sptr holding the simulation has a nullptr by default.

damienmarchal committed:
[all] Update the codebase to take into account the change in Simulation.h.

damienmarchal committed:
[SofaPython3] FIX SofaRuntime.

marques-bruno committed:
[pybind] FIX Base methods access in Node

marques-bruno committed:
[pybind11] ADD: few methods for Binding Vec

marques-bruno committed:
[pybind] ADD: Vector class bindings with template specializations for ctors & sets

marques-bruno committed:
[pybind] ADD: more bindings for the Vector class

marques-bruno committed:
[pybind] ADD more Vector bindings

marques-bruno committed:
[pybind] ADD first binding draft for Matrices & Quaternions

marques-bruno committed:
[pybind] ADD first binding draft for Matrices & Quaternions

marques-bruno committed:
[pybind11][SofaTypes]: ADD more Matrix / Quat / Vector bindings + tests

marques-bruno committed:
[pybind11][SofaTypes] ADD tests and fix vector, mat & quat bindins

damienmarchal committed:
[SofaPython3/Sofa] Add few methods & PythonController.

damienmarchal committed:
[SofaPython3/Sofa] Add missing PythonController.cpp

damienmarchal committed:
Ha ba si :)

damienmarchal committed:
[SofaPython3/Sofa] Add PythonController tests v2.

damienmarchal committed:
[SofaPython3/Sofa] Add missing return in PythonController.

damienmarchal committed:
[SofaPython3/binding] WIP in PythonController.cpp

damienmarchal committed:
[SofaPython3/Sofa] WIP ScriptController.cpp.

marques-bruno committed:
[pybind11][Sofa] ADD a working version of the PythonController, through a wrapping class

marques-bruno committed:
[pybind11][SofaTypes] ADD: few extra tests

marques-bruno committed:
Merge branch 'sofa_refactorSofaPython' of https://github.com/SofaDefrost/sofa into sofa_refactorSofaPython

damienmarchal committed:
[SofaPython3] Force python version to be 3.6m to fix link issue (WIP)

damienmarchal committed:
[SofaPython3] There is lot of problem in cmakelists (WIP)

damienmarchal committed:
[SofaPython3] Make a working environment & loader.

damienmarchal committed:
[SofaPython3] ADD a dev_getANode() method in SofaRuntime for dev purpose.

damienmarchal committed:
[SofaPython3/bindings/Sofa] A less invasive version of PythonController.

damienmarchal committed:
[SofaPython3] WIP still stuck on PythonController.

marques-bruno committed:
[pybind11] finalizing vector's bindings & tests

damienmarchal committed:
[SofaKernel] Move some function into StringUtils.

damienmarchal committed:
[SofaPython3] Up.

marques-bruno committed:
Merge branch 'sofa_refactorSofaPython' of https://github.com/SofaDefrost/sofa into sofa_refactorSofaPython

damienmarchal committed:
[SofaKernel & extlibs] Experimental change.

To allow the intrusive ptr to be manipulated from outside.
To prevent gtest to fetch the python version.

damienmarchal committed:
Merge branch 'sofa_refactorSofaPython' of github.com:SofaDefrost/sofa into sofa_refactorSofaPython

damienmarchal committed:
[SofaKernel] Move code from .h to a .cpp

damienmarchal committed:
[SofaPython3] Fix CMakeLists.txt for SofaRuntime.

damienmarchal committed:
[SofaKernel] Clean SimpleApi

- Remove the pass by reference to prefer full smart pointer passing.
This choice is guided by the fact that with reference only it is not
possible to reconstruct a temporary smart pointer out of a pointer.

- Remove some experimental code that was not supposed to be in Sofa
and for which there is no chance to be deployed massively.


Signed-off-by: Damien Marchal <damien.marchal@univ-lille1.fr>

damienmarchal committed:
[SofaPython3] Add a drafted implementation of the protocol buffer.

This protocol allows python to echange data in a flexible,
yet efficient way, pieces of c++ code.

damienmarchal committed:
[SofaPython3] Stop hard coding the SOFA_ROOT in runSimu.py.

damienmarchal committed:
[SofaPython3] ADD news bindings in Binding_Node.

damienmarchal committed:
[SofaPython] Add Binding_DataArray.cpp/h

This is not the thing to do as this can be generated from the .pyx.

damienmarchal committed:
[SofaPython3] Rename moduleAddDataContainer.

damienmarchal committed:
[SofaPython3] Minor work in the testing .py

damienmarchal committed:
[SofaTypes] Refactor Binding_Vec (WIP, some tests are failing)

damienmarchal committed:
[SofaPython] Port features un Base*

damienmarchal committed:
[SofaPython3] Activate SofaTypes

damienmarchal committed:
[SofaPython3] Fix the vector __imul__ & __idiv__

In python, the imul & idiv are operator that modify things in place
but they needs to return the value.

damienmarchal committed:
[SofaPython3] Fix Binding_Vec __str__ and tests.

damienmarchal committed:
[SofaPython3] Minore cleaning of Binding_mat.cpp

Signed-off-by: Damien Marchal <damien.marchal@univ-lille1.fr>

damienmarchal committed:
[SofaPython3] Binding_vec, remove useless code.

damienmarchal committed:
[SofaKernel] ADD Dev message in Node.cpp

damienmarchal committed:
[SofaPython3] ADD tests & features in the Sofa binding.

There is now a __data__ attribute in object so we can
access the data in a way similar to classical attributes
in python that are in __dict__.

This __data__ is iterable/mapping.

damienmarchal committed:
[SofaPython3] Remove useless dependency to SofaTest in CMakeLists.txt

damienmarchal committed:
[SofaPython3] Try to stabilize the PythonEnvironment (but this is hard)

damienmarchal committed:
[SofaPython3] ADD a getTestName function to PythonTest

This allows to have the filename exposed into the test name.
The consequence is that it is possible to refine the test running
with --gtest_filter=*FileName*

damienmarchal committed:
[SofaPython3] WIP on data access in binding/Sofa

The general idea is to have a buffer protocol & numpy array raw acces
to objects.

damienmarchal committed:
[SofaPython3] There is still problem while releasing the environment (crash).

damienmarchal committed:
[SofaPython3] Replace the use of Node::create with DAGNode::create in SceneLoader3.

The problem with Node::create is that it use the instance of
Simulation that is loaded (tree vs graph). Which is not really
consistant. It is much better that our API only work with DAG.
And if needed can be mapped to tree structure.

Eg:
  from Sofa.simulation import Node it is actually sub-loading
Sofa.Simulation.DAG.Node


This also allow to write explicit call as like this:
  from Sofa.Simulation.DAG import Node
vs
  from Sofa.Simulation.Tree import Node

damienmarchal committed:
[SofaKernel] FIX some problemn in TestMessageHandler.

- remove the use of a static instance as the instance was
  removed too early in the de-allocation process causing crahs.
  Now there is an object that is not de-allocated (I'm not sure
  this is a good thing but at least this does not crash).

- in SofaGTestMain some lines seems to be at wrong location.

damienmarchal committed:
[SofaPython3] Reorganise Sofa to mimick a bit more the Sofa oranization.

Now Sofa is amodule with Submodules:
Sofa.Core
Sofa.Simulation
Sofa.Types (to do)

The object that can be instantiated are exposed in "Sofa", the other needs
their full path.

I don't know if it is better or not. Please give feedback in case you disagree.

damienmarchal committed:
[SofaPython3/binding/Sofa] Reorganisation of the source code.

(see previous commit)

damienmarchal committed:
[SofaPython3] Add SofaPython3_SofaConfig.cmake.in for inclusion in find_apckage.

damienmarchal committed:
[SofaKernel] add std::string version instead of const char*

damienmarchal committed:
[SofaKernel] SimpleApi change to support BaseObjectDescription

damienmarchal committed:
[SofaPython3] fix test in createObject.

The underlying problem is how to we report error from Sofa to python.
Ideally , exception should be used.

damienmarchal committed:
[SofaKernel] CLEAN remove the duplicated method and use default args BaseData.

damienmarchal committed:
[SOfaKernel] Replace findData in Base.cpp with a simpler method.

damienmarchal committed:
[runSofa] ADD a break in a broken swithc/case

damienmarchal committed:
[SofaPython3] WIP Data access on steroïd (array + object cache)

1) I have implemented a cache to avoid converting multiple time
the same object. This cache is greatly speeds up every operations involving
the data. A serious cache policy (LRU, etc..) needs to be implemented.

2) There is also the results from my experiments on accessing the
data as numpy.array and how this can be done without breaking
the DataEngines semantics.

damienmarchal commented:
Hi all,

I'm hesitating between the following two design:

```__getattr__``` returns a DataContainer object implementing lot of operations allowing:
```python
root = Sofa.Node("root")
obj = root.createObject("MechanicalObject", name="test", position=aList)

np = numpy.zeros((100,3))

p.position = [[1.0,2.0,3.0]]
p.position = np
p.position += [1.0,2.0,3.0]
p.position += np
p.position += Vec3d((0.,0.0,0.0))

## To get a numpy array view out the DataContainer it is needed to request it like that:
with p.position.array() as wa:  
        wa += np 
        wa *= np
          # Other numpy.array operation
```
This syntax guarantee that the beginEditPtr/endEditPtr is correctly called when the data are written. The drawback is that I have to re+implement a lot of operation on the DataContainer object (add,mul, iadd, imul, etc..etc...)which is annoying.

The other approach is much lighter in term of code as it is directly a numpy.array out of the ```__getattr__``` function. This allows to do
```python
np = numpy.zeros((100,3))

p.position = [[1.0,2.0,3.0]]
p.position = np
p.position += [1.0,2.0,3.0]
p.position += np
p.position += Vec3d((0.,0.0,0.0))

## To get have correct management of begin/end... you should do like that:
with p.WriteAccessor("position") as wa:
          wa += np 
          # Other numpy.array operation
```
The problem with this version is that doing:
```
d = p.position
d += np
```
Will change the data through numpy code without incrementing the data counter which violates the expected semantic of the DataEngine mechanismes.  I have tried different solution in the last two days but cannot find one. 

So any comment and help is welcomed :)

damienmarchal committed:
[SofaPython3] WIP about DataContainer.

This version sounds much better and feature most of what we need.
There is drafted implementation of the different data access syntax,
including one based on the contextmanager.

Have alook at the dataaccess.py for examples.

damienmarchal committed:
[SofaPython3] ADD to DataAsContainer a 'tolist' method.

damienmarchal committed:
[SofaPython3] ADD bench_datacontainer.py so that we have baseline for optimization.

damienmarchal committed:
[SofaPython3] FIX several problems and memory leaks in how we are caching data containers.

The general problem was that the caching system was holding reference
to destroyed object. So it was necessary to hold a smart pointer to the
data sources as long as they are in the cache.

Open question: do we really need such cache instead of storing the data in
a different object ?

Signed-off-by: Damien Marchal <damien.marchal@univ-lille1.fr>

damienmarchal committed:
[SofaPython3] ADD 'children' and 'parents' properties to Node

damienmarchal committed:
[SofaPython3] FIX the tests.

damienmarchal committed:
[SofaPython3] ADD the wrap-around types to decorate data with their right type.

Currently the data are exposed as "DataContainer" in python. The reason is that it
allow to offer an unified way to manipulate all the data fields as well as insuring
that the beginEdit/endEdit pair are correctly handled. The user can also get
read or write access on the data field through numpy.view of the data.

Now, it is often convenient to manipulate our data with more concrete type (eg: Vec4,
Rigid, Frame, RGBAColor). One possibility would be to implement those data types
in the binding, the approach I tried in this commit is different as the concrete
types are implemented as a kind of façade over the numpy.array exposed by the DataContainer.
This is what is implemented in this commit.

The general idea is to wrap subtypes around a compatible DataContainer.
For the user point of view this look like that:
color = RGBAColor(obj.showColor)
v = Vec4(obj.position)
rigid = Rigid(obj.position)

Then it become possible to manipulate the DataContainer as if it
is a color (so using dedicated methods).
color.r()
color.hsv()

The previous syntax is in fact equivalent to write:
RGBAColor(obj.showColor.toarray())

So it is clear that because of the use of the toarray() function we are operating
in read-only mode. So do things like that:
color.r = 40 or color.setFromHSV(1,2,3)
will generate an exception.

In order to get write access it is needed to use the provided context manager in
the following way:
Eg:
with obj.showColor.writeable(RGBAColor) as color:
	color.r = 40

when the context left, the color object is still accessible but it returns back to
a read-only state.

damienmarchal committed:
[SofaPython3] Unify the implementation of WriteAccessor

It is now renamed into DataContainerContextManager in the python side.
I also clean several files.

damienmarchal committed:
[SofaPython3] ADD new benchmarks.

damienmarchal committed:
[SofaPython3] Remove some compilation warnings.

damienmarchal committed:
[SofaPython3] CLEAN & REFACTOR.

damienmarchal committed:
[SofaPython3] Add new function to Sofa module

damienmarchal committed:
[SofaPython3] Refactor PythonController to make it more consistant

damienmarchal committed:
[SofaPython3] ADD A basic Sofa.ForceField that can be overriden in python

This is a WIP but as the ForceField is sending the object as Data<> and
we have a nice API for the data fields can be do interesting things out of this.

damienmarchal commented:
I have added a bunch of new functions and a draft on how to implement force field in python. 

It taste like that:
```python
class MyForceField(Sofa.Core.BaseForceField):
    def __init__(self, *args, **kwargs):
        Sofa.Core.BaseForceField.__init__(self, *args, **kwargs)

    def addForce(self, m, f, x, v):
        n = numpy.ones((1000,1))
        f = x * 2 + n
    
class Test(unittest.TestCase):
    def test_animation(self):
        node = Sofa.Node("root")
        node.addObject("DefaultAnimationLoop", name="loop")
        node.addObject("EulerImplicit")
        node.addObject("CGLinearSolver")
        object1 = node.addChild("object1")
        c = object1.addObject("MechanicalObject", position=[0,0,0]*1000)
        d = object1.addObject(MyForceField("customFF"))

        SingleSimulation.init(node)
        for i in range(10):
            SingleSimulation.animate(node, 0.01)
```

jnbrunet commented on SofaKernel/framework/sofa/simulation/Simulation.cpp:
```diff
@@ -490,6 +473,25 @@ void Simulation::unload(Node::SPtr root)
+/// simulations at the same time (e.g. DAGSimulation + TreeSimulation)
+/// but it does not sound like a huge limitation
+/// The (unique) simulation which controls the scene
+Simulation::SPtr theSimulation  {nullptr};
```
I know this isn't related to this PR, but I need to say it :-P

I don't think a singleton here is very elegant, and can lead to overuse of singletons and losing track of what is created and when it is destroyed. I believe it should be up to the application to hold its simulation...or simulations (one should have the possibility to create many simulations all independent from each other without having the impression of not following sofa's principles).

In the case of runSofa, it could create the simulation's instance, hold its reference, and delete it once the GUI or main loop ends. But it should not be the place of the kernel (this file) to do that.

damienmarchal replied:
Thank for sharing, I think there is since years a general agreement on that feel free to open an issue or a PR to fix :) 

In the meantime, to mitigate the problem in the context of SofaPython3 the design I have taken is the following:
- Sofa.Core.Simulation (only a DAG simulation)
- Sofa.Core.Node (actually only a DAG node I see no reason why tree node are needed)
These never (I need to double check that :)) manipulates the singleton. 

Manipulation of the 'application' singleton is done via the  Sofa.Runtime module. For the moment to avoid extra work this one contains only Sofa.Runtime.SingleSimulation to manipulate this singleton:
```python
   Sofa.Runtime.SingleSimulation.setSimulation(Sofa.Core.Simulation())

   # For the future (ie a multi simulation graph application)
   Sofa.Runtime.Simulations.addSimulation(Sofa.Core.Simulation()) 
   Sofa.Runtime.Simulations.removeSimulation(Sofa.Core.Simulation())
```

I hope that with this design if one day someone refactor the simulation singleton in Sofa then the python will still looks ok.

damienmarchal committed:
[SofaKernel] REMOVE commented code un MutliVecId.h

damienmarchal committed:
[SofaPython3] A first version of ForceField that works.

damienmarchal committed:
[SofaPython3] FIX bug in PythonEnvironment

damienmarchal committed:
[SofaPython3] ADD the adding of data field directly from the syntax a.toto = 5.0

damienmarchal committed:
[SofaPython3] WIP on making force field in python.

damienmarchal committed:
[SofaPython] Example of a rest shape spring forcefield in python.

When using only numpy call (no real python in it) it not very much slower
than c++ code.

damienmarchal committed:
[SofaPython3] Add a version that use autodiff to compute the jacobian used for addDForce.

To me this is fundamentally ill-designed.

damienmarchal committed:
[SofaPython3] CLEAN of the AutoDiff example

damienmarchal committed:
[SofaPython3] Add '#coding: utf8' in python files.

damienmarchal committed:
[SofaPython3] ADD Sofa.Helper submodule, msg_* API and corresponding tests.

damienmarchal committed:
[SofaPython3] FIX tests, Switch to wrap-around design, remove deadfiles.

damienmarchal committed:
[SofaPython3] CMakeLists.txt fixes compilation issue related to remove of deadcode.

damienmarchal committed:
[SofaPython3] Post-meeting changes.

damienmarchal committed:
[SofaPython3] Add setToDict, setToData, getFromDict, getFromData method in Base

The idea is to have general symetry on the binding where we have:
c.__dict__
c.setToDict("aName", value)
c.getFromDict("aName")

c.__data__
c.setToData("aName", value)
c.getFromData("aName")


Signed-off-by: Damien Marchal <damien.marchal@univ-lille1.fr>

JeremieA commented on SofaKernel/modules/SofaSimulationGraph/DAGNode.cpp:
```diff
@@ -330,12 +330,13 @@ void DAGNode::getObjects(const sofa::core::objectmodel::ClassInfo& class_info, G
+            break;
         case SearchParents:
         {
+            this->getLocalObjects( class_info, container, tags ); // add locals then SearchParents
```
I came upon this while randomly looking at open merge requests: this change does look wrong to me.
The original code was using the fact that SearchUp is supposed to search for local objects and then within parents, while SearchParents should only search within parent. Without the break, both cases could share the same code for the shared part.
The changed code is no longer looking in parents for SearchUp while it should, and is searching for local objects for SearchParents while it should not...
I suspect this was done because of the gcc warning about implicit-fallthrough. There are several ways to silence that warning, depending on the c++ standard that should be supported (see [here](http://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-fallthrough_003d)). The easiest is to add a comment `/* FALLTHRU */` above the `case`.

damienmarchal replied:
You are defintely right, thanks for reporting.

damienmarchal commented:
Hi all, 

I close this PR because this plugin is now here: https://github.com/SofaDefrost/plugin.SofaPython3
